sharding/scheme.go

```diff
@@ -22,73 +22,59 @@ package sharding
 
 import (
 	"errors"
+	"math"
 )
 
 var (
-	// ErrToLessThanFrom returned when to is less than from
-	ErrToLessThanFrom = errors.New("to is less than from")
 	// ErrNoShards returned when shard set is empty
 	ErrNoShards = errors.New("empty shard set")
 	// ErrDuplicateShards returned when shard set is empty
 	ErrDuplicateShards = errors.New("duplicate shards")
 )
 
-type shardScheme struct {
+type shardSet struct {
 	shards []uint32
 	fn   HashFn
 }
 
-// NewShardSchemeFromRange creates a new sharding scheme, from and to are inclusive
-func NewShardSchemeFromRange(from, to uint32, fn HashFn) (ShardScheme, error) {
-	if to < from {
-		return nil, ErrToLessThanFrom
-	}
-	var shards []uint32
-	for i := from; i <= to; i++ {
-		shards = append(shards, i)
-	}
-	return NewShardScheme(shards, fn)
-}
-
-// NewShardScheme creates a new sharding scheme with a set of shards
-func NewShardScheme(shards []uint32, fn HashFn) (ShardScheme, error) {
+// NewShardSet creates a new sharding scheme with a set of shards
+func NewShardSet(shards []uint32, fn HashFn) (ShardSet, error) {
 	if err := validateShards(shards); err != nil {
 		return nil, err
 	}
-	return &shardScheme{shards, fn}, nil
+	return &shardSet{shards, fn}, nil
 }
 
-func (s *shardScheme) Shard(identifer string) uint32 {
+func (s *shardSet) Shard(identifer string) uint32 {
 	return s.fn(identifer)
 }
 
-func (s *shardScheme) CreateSet(shards []uint32) (ShardSet, error) {
-	if err := validateShards(shards); err != nil {
-		return nil, err
-	}
-	return NewShardSet(shards, s), nil
-}
-
-func (s *shardScheme) All() ShardSet {
-	return NewShardSet(s.shards, s)
-}
-
-type shardSet struct {
-	shards []uint32
-	scheme ShardScheme
+func (s *shardSet) Shards() []uint32 {
+	return s.shards[:]
 }
 
-// NewShardSet creates a new shard set
-func NewShardSet(shards []uint32, scheme ShardScheme) ShardSet {
-	return &shardSet{shards, scheme}
+func (s *shardSet) Min() uint32 {
+	min := uint32(math.MaxUint32)
```

Comment:
can do, but Min() is not being used now and Max() is only used when creating a DB, not sure how much value added from the caching